chronos/threadsync.nim

```diff
@@ -0,0 +1,416 @@
+      var data = 0'u64
+      let res = handleEintr(recv(SocketHandle(fd),
+                                 addr data, sizeof(uint64), MSG_PEEK))
+      if res == sizeof(uint64):
+        true
+      else:
+        false
+
+  func toTimeval(a: Duration): Timeval =
+    ## Convert Duration ``a`` to ``Timeval`` object.
+    let nanos = a.nanoseconds
+    let m = nanos mod Second.nanoseconds()
+    Timeval(
+      tv_sec: Time(nanos div Second.nanoseconds()),
+      tv_usec: Suseconds(m div Microsecond.nanoseconds())
+    )
+
+  proc waitReady(fd: cint, kind: WaitKind,
+                 timeout: Duration): Result[bool, OSErrorCode] =
+    var
+      tv: Timeval
+      fdset =
+        block:
+          var res: TFdSet
+          FD_ZERO(res)
+          FD_SET(SocketHandle(fd), res)
+          res
+
+    let
+      ptv =
+        if not(timeout.isInfinite()):
+          tv = timeout.toTimeval()
+          addr tv
+        else:
+          nil
+      nfd = cint(fd) + 1
+      res =
+        case kind
+        of WaitKind.Read:
+          handleEintr(select(nfd, addr fdset, nil, nil, ptv))
+        of WaitKind.Write:
+          handleEintr(select(nfd, nil, addr fdset, nil, ptv))
+
+    if res > 0:
+      ok(true)
+    elif res == 0:
+      ok(false)
+    else:
+      err(osLastError())
+
+  proc safeUnregisterAndCloseFd(fd: AsyncFD): Result[void, OSErrorCode] =
+    let loop = getThreadDispatcher()
+    if loop.contains(fd):
+      ? unregister2(fd)
+    if closeFd(cint(fd)) != 0:
+      err(osLastError())
+    else:
+      ok()
+
+proc close*(signal: ThreadSignalPtr): Result[void, string] =
+  ## Close ThreadSignal object and free all the resources.
+  defer: deallocShared(signal)
+  when defined(windows):
+    # We do not need to perform unregistering on Windows, we can only close it.
+    if closeHandle(signal[].event) == 0'u32:
+      return err(osErrorMsg(osLastError()))
+  elif defined(linux):
+    let res = safeUnregisterAndCloseFd(signal[].efd)
+    if res.isErr():
+      return err(osErrorMsg(res.error))
+  else:
+    let res1 = safeUnregisterAndCloseFd(signal[].rfd)
+    let res2 = safeUnregisterAndCloseFd(signal[].wfd)
+    if res1.isErr(): return err(osErrorMsg(res1.error))
+    if res2.isErr(): return err(osErrorMsg(res2.error))
+  ok()
+
+proc fireSync*(signal: ThreadSignalPtr,
+               timeout = InfiniteDuration): Result[bool, string] =
+  ## Set state of ``signal`` to signaled state in blocking way.
+  ##
+  ## Returns ``false`` if signal was not signalled in time, and ``true``
+  ## if operation was successful.
+  when defined(windows):
+    if setEvent(signal[].event) == 0'u32:
+      return err(osErrorMsg(osLastError()))
+    ok(true)
+  else:
+    let
+      eventFd =
+        when defined(linux):
+          cint(signal[].efd)
+        else:
+          cint(signal[].wfd)
+      checkFd =
+        when defined(linux):
+          cint(signal[].efd)
+        else:
+          cint(signal[].rfd)
+
+    if checkBusy(checkFd):
```

Comment:
is it thread-safe to skip the writing here? if another thread reads just after this check, it might reset the conditions for the "writer" to fire again